### ISSUE TITLE
Too Slow

### ISSUE THREAD

**Issue by Waterman178** [Cisco-Talos/pyrebox#76]
I successfully installed win7x64 system.But because it supports up to 2047MB of memory.And it seems to only support single-core cpu,so the system is very slow.I open the task manager,I found that although the memory only uses 450MB, the cpu is 100%.So slow, you can't debug any process at all.I am using a debian-based operating system created by vmware.In the debian system, I created the win7x64 system using qemu.So if I use the qemu to create a win7x64 system directly on the real machine, can it be faster?I feel that even if it is running on a real machine, it is still very slow because it only supports single-core cpu.Is there any way to make the system run faster?

Host:debian  created by vmware
Guest:win7x64
Command:.
/pyrebox-x86_64 -m 2047 -monitor stdio -usb -drive file=images/Windows7x64.qcow2,index=0,media=disk,format=qcow2,cache=unsafe -cdrom images/Windows7x64.iso -boot d

**Comment by xabiugarte** [comment]
Hi,

If you run PyREBox on the host system instead of a VM, it might run slightly faster, but the difference should not be too big. QEMU allows to emulate more than one virtual CPU, by using the -smp command line option. In the past, all vCPUs were emulated in a single host thread, so there was no difference (in terms of performance) between emulating one or several vCPUs.

Recently (december 2018), in QEMU version 3.1.0,  they have introduced the support for MTTCG for x86/64, which means that vCPUs can be emulated in different threads. This will bring a very good performance improvement.

I am currently working on the integration of this new version into PyREBox. This change is not trivial and will require some thorough testing so I will not be releasing it until it is ready.

So, in summary:

QEMU (in whole system emulation mode) and as a consequence PyREBox, are inherently slower than a real system or an hypervisor, because they are based on emulation. MTTCG will bring a significant improvement, but it will still be slower than native execution.

Please reopen the issue if you have additional related questions. 

Yours,

Xabier

**Comment by Waterman178** [comment]
> Hi,
> 
> If you run PyREBox on the host system instead of a VM, it might run slightly faster, but the difference should not be too big. QEMU allows to emulate more than one virtual CPU, by using the -smp command line option. In the past, all vCPUs were emulated in a single host thread, so there was no difference (in terms of performance) between emulating one or several vCPUs.
> 
> Recently (december 2018), in QEMU version 3.1.0, they have introduced the support for MTTCG for x86/64, which means that vCPUs can be emulated in different threads. This will bring a very good performance improvement.
> 
> I am currently working on the integration of this new version into PyREBox. This change is not trivial and will require some thorough testing so I will not be releasing it until it is ready.
> 
> So, in summary:
> 
> QEMU (in whole system emulation mode) and as a consequence PyREBox, are inherently slower than a real system or an hypervisor, because they are based on emulation. MTTCG will bring a significant improvement, but it will still be slower than native execution.
> 
> Please reopen the issue if you have additional related questions.
> 
> Yours,
> 
> Xabier

Thank you very much for your contribution.It seems that I can only wait for your final release.I want to know if there is a similar debugging system?It is not necessary to simulate the entire system.I only need it to simulate the execution of the target process.Or is there any related project?I can refer to writing a similar simulated sandbox.
